src/main/scala/state/Terms.scala

```diff
@@ -2295,48 +2300,139 @@ object PredicateTrigger extends PreciseCondFlyweightFactory[(String, Term, Seq[T
 
 /* Magic wands */
 
-class MagicWandSnapshot(val abstractLhs: Term, val rhsSnapshot: Term) extends Combine(abstractLhs, rhsSnapshot) {
+abstract class MagicWandSnapshot(val abstractLhs: Term, val rhsSnapshot: Term) extends Term {
   utils.assertSort(abstractLhs, "abstract lhs", sorts.Snap)
-  utils.assertSort(rhsSnapshot, "rhs", sorts.Snap)
+  utils.assertSort(rhsSnapshot, "rhs snapshot", sorts.Snap)
+}
 
-  override lazy val toString = s"wandSnap(lhs = $abstractLhs, rhs = $rhsSnapshot)"
+/**
+ * Snapshot for a magic wand. It represents a function which can be used only once
+ * by equating a snapshot from consuming the wand's LHS with `abstractLhs`.
+ *
+ * For a snapshot which can be applied multiple times see [[MagicWandSnapFunction]].
+ *
+ * @param abstractLhs The abstract snapshot which acts as a placeholder in the
+ *                    `rhsSnapshot` for the values when applying the magic wand.
+ * @param rhsSnapshot Snapshot which can be used when producing the wand's RHS.
+ */
+class MagicWandSnapSingleton(override val abstractLhs: Term, override val rhsSnapshot: Term)
+  extends MagicWandSnapshot(abstractLhs, rhsSnapshot)
+    with ConditionalFlyweightBinaryOp[MagicWandSnapSingleton] {
 
-  def merge(other: MagicWandSnapshot, branchConditions: Stack[Term]): MagicWandSnapshot = {
-    assert(this.abstractLhs == other.abstractLhs)
-    val condition = And(branchConditions)
-    MagicWandSnapshot(this.abstractLhs, if (this.rhsSnapshot == other.rhsSnapshot)
-      this.rhsSnapshot
-    else
-      Ite(condition, other.rhsSnapshot, this.rhsSnapshot))
-  }
+  override lazy val toString: String = s"wandSnap(lhs = $abstractLhs, rhs = $rhsSnapshot)"
+
+  /* ConditionalFlyweightBinaryOp members */
+
+  override def sort: Sort = sorts.Snap
+
+  override def p0: Term = abstractLhs
+
+  override def p1: Term = rhsSnapshot
 }
 
-object MagicWandSnapshot  {
+object MagicWandSnapSingleton {
+  var pool = new TrieMap[(Term, Term), MagicWandSnapSingleton]()
```

Comment:
Is there a reason you have to declare the pool yourself here and can't inherit it from ConditionalFlyweightFactory?